README.md

Incomplete data is a common challenge when analyzing real-world data.
Another challenge is choosing from many available multiple imputation
approaches to fill-in the missing values. Marbach (2021) suggests
adopting the imputation approach that generates a density of imputed
values most similar to those of the observed values for an incomplete
variable after balancing all other covariates. The package facilitates
the computation of discrepancy statistics summarizing differences
between the density of imputed and observed values and the construction
of weights to balance covariates.

### Installation

``` r
remotes:::install_github("sumtxt/missDiag")
```

To install the `remotes` package use `install.packages("remotes")`.

### Requirements

The package can be installed with few dependencies but to construct
weights the package relies on either the
[ebal](https://cran.r-project.org/web/packages/ebal/index.html) package
(Hainmueller, 2012) or the
[sbw](https://cran.r-project.org/web/packages/sbw/index.html) package
(Zubizarreta 2015). These two packages have be installed by the users
separately.

To install these packages from CRAN use:

``` r
remotes::install_cran("ebal")
remotes::install_cran("sbw")
```

### Usage

The package comes with some example data from the 2008 American National
Election Study (`anes08`) in which, for example, about 11% of the values
in the vote-choice variable are missing. The data have been imputed five
times using random value imputation (output data: `anes08_rng`) and
predictive mean matching imputation (`anes08_pmm`). We use the
`missDiag` package to corroborate that predictive mean matching produces
a density of imputed values that is more similar to the (reweighted)
density of observed values and should therefore be preferred over random
value imputation. We use entropy-balancing weights from the `ebal`
package as their computation is faster but stable weights from the `sbw`
package are preferable as their variance is smaller.

``` r
library(tidyverse)
#> ── Attaching packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.1 ──
#> ✔ ggplot2 3.3.5     ✔ purrr   0.3.4
#> ✔ tibble  3.1.3     ✔ dplyr   1.0.7
#> ✔ tidyr   1.1.3     ✔ stringr 1.4.0
#> ✔ readr   2.0.0     ✔ forcats 0.5.1
#> ── Conflicts ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
#> ✖ dplyr::filter() masks stats::filter()
#> ✖ dplyr::lag()    masks stats::lag()
library(missDiag)

diag_rng <- missDiag( 
 original=anes08, 
 imputed=anes08_rng, 
 adjust = 'ebal',
 formula = vote ~ .)

diag_pmm <- missDiag( 
 original=anes08, 
 imputed=anes08_pmm, 
 adjust = 'ebal',
 formula = vote ~ .)
```

Averaging across the five multiple imputed datasets, we compare the
standardized mean differences (SMD) for each of the vote choice
categories and find that predictive mean matching does much better than
random value imputation. The standardized mean differences are, on
average, lower for predictive mean matching.

``` r
diag_rng %>% group_by(vname) %>% 
    summarize( smd=mean(diff_adj) )
#> # A tibble: 3 × 2
#>   vname                 smd
#>   <chr>               <dbl>
#> 1 vote_McCain        0.0621
#> 2 vote_No vote/Other 0.0731
#> 3 vote_Obama         0.0577

diag_pmm %>% group_by(vname) %>% 
    summarize( smd=mean(diff_adj) )
#> # A tibble: 3 × 2
#>   vname                 smd
#>   <chr>               <dbl>
#> 1 vote_McCain        0.0562
#> 2 vote_No vote/Other 0.0417
#> 3 vote_Obama         0.0512
```

### References

Moritz Marbach. 2021. [Choosing Imputation
Models](https://arxiv.org/abs/2107.05427), Political Analysis
(forthcoming).

José R Zubizarreta. 2015. [Stable Weights that Balance Covariates for
Estimation with Incomplete Outcome
Data](https://doi.org/10.1080/01621459.2015.1023805), Journal of the
American Statistical Association, 110(511): 910-922.

Jens Hainmueller. 2012. [Entropy Balancing for Causal Effects: A
Multivariate Reweighting Method to Produce Balanced Samples in
Observational Studies](https://doi.org/10.1093/pan/mpr025), Political
Analysis 20(1): 25–46.


NEWS.md
# missDiag

# missDiag 1.0.1

* First CRAN release

R/param.R
#' Construct parameter list for ebalance
#' 
#' @param coefs starting values for model coefficients.
#' @param max.iterations maximum number of iterations.
#' @param base.weight vector of base weights.
#' @param constraint.tolerance tolerance level.
#' @param norm.constant An optional normalizing constant.
#' @param trim trim weights via \code{\link[ebal]{ebalance.trim}}
#' @param max.weight Target for the ratio of the maximum to mean weight.
#' @param min.weight Target for the ratio of the minimum to mean weight.
#' @param max.trim.iterations Maximum number of trimming iterations.
#' @param max.weight.increment Increment for iterative trimming of the 
#'  ratio of the maximum to mean weight.
#' @param min.weight.increment Increment for iterative trimming of the 
#'  ratio of the minimum to mean weight.
#' 
#' @details 
#' For more information about these parameters, see 
#' [ebal::ebalance()] and [ebal::ebalance.trim()].
#' 
#' @return A \code{list} of parameters that can be passed to [missDiag()].
#' 
#' @export
param_ebal <- function(
  coefs = NULL,
  max.iterations = 200,
  base.weight = NULL,
  constraint.tolerance = 1,
  norm.constant = NULL,
  trim = FALSE,
  max.weight = NULL,
  min.weight = 0,
  max.trim.iterations = 200,
  max.weight.increment = 0.92,
  min.weight.increment = 1.08){
  return(as.list(environment()))
  }

#' Construct parameter list for sbw
#' 
#' @param bal_alg use tuning algorithm as in Wang and Zubizarreta (2020)? 
#' @param bal_tol tolerance level.
#' @param bal_std tolerances adjustment.
#' @param bal_gri grid of values for the tuning algorithm.
#' @param bal_sam number of replicates to be used by the tuning algorithm.
#' @param sol_nam solver name. Either "cplex", "gurobi", "mosek", 
#'  "pogs", or "quadprog".
#' 
#' 
#' @details 
#' For more information about these parameters, see [sbw::sbw()].
#' 
#' @return A \code{list} of parameters that can be passed to [missDiag()].
#' 
#' @export
param_sbw <- function(
    bal_alg = TRUE,
    bal_tol = 0,
    bal_std = "group",
    bal_gri = c(1e-04,0.001,0.002,0.005,0.01,0.02,0.05,0.1),
    bal_sam = 1000,
    sol_nam = "quadprog"){

  return(as.list(environment()))  
  }

#' Construct parameter list for cobalt
#' 
#' @param stats character; which statistic(s) should be reported.
#' @param continuous whether mean differences for continuous variables 
#'  should be standardized ("std") or raw ("raw").
#' @param binary whether mean differences for binary variables (i.e., 
#'  difference in proportion) should be standardized ("std") or raw ("raw").
#' @param abs logical; whether displayed balance statistics 
#'  should be in absolute value or not.
#' @param s.d.denom character; how the denominator for standardized 
#'  mean differences should be calculated, if requested.
#' 
#' 
#' @details 
#' For more information about these parameters, see 
#'  [cobalt::bal.tab()].
#' 
#' @return A \code{list} of parameters that can be passed to [missDiag()].
#' 
#' @export
param_cobalt <- function(
  stats = c("m", "v", "ks", "ovl"), 
  continuous = "std", 
  binary = "std", 
  abs = TRUE,
  s.d.denom = 'pooled'){
  return(as.list(environment()))  
  }




R/utils.R
expand_rhs_dot <- function(f,all_vars){

  if( sum(all.vars(f) == ".") == 0 ){
  
    return(f)
  
  } else {

    dv <- all.vars(f)[1]
    iv <- paste(all_vars, collapse="+")
    f <- as.formula(paste0(dv, "~", iv))
    return(f)

  }
  }

#' @import Formula
construct_formulas <- function(f, vars){

  if(is.list(f)){

    formulas <- lapply(f, as.formula)
    formulas <- lapply(f, expand_rhs_dot, all_vars=vars)

  } else {

    f <- as.Formula(f)
    K <- length(f)[1]

    formulas <- list()

    for(i in 1:K){
      
       nf <- formula(as.Formula(f),lhs=i)
       nf <- expand_rhs_dot(nf, all_vars=vars)
       nf <- update.formula(nf, 
          paste(" . ~ . -", all.vars(nf)[1]),
          evaluate=FALSE)

      formulas[[i]] <- nf

      }

  }

  return(formulas)
  }


# construct_formulas <- function(f){
# 
#   if(is.list(f)){
# 
#     formulas <- lapply(f, as.formula)
# 
#   } else {
# 
#     f <- as.Formula(f)
#     K <- length(f)[1]
# 
#     formulas <- list()
#     for(i in 1:K){
#        nf <- formula(as.Formula(f),lhs=i)
#        nf <- update.formula(nf, 
#           paste(" . ~ . -", all.vars(nf)[1]),
#           evaluate=FALSE)
#       formulas[[i]] <- nf
#       }
# 
#   }
# 
#   return(formulas)
#   }


formula_string <- function(f){
  f <- paste(deparse(f),collapse="")
  f <- gsub("\\s+", " ", f)
  return(f)
  }

construct_x <- function(df,formula,m,scale,verbose){

  X <- model.matrix(formula,df)
  X <- X[,colnames(X)!="(Intercept)", drop=FALSE]
  if(scale) X <- scale(X)

  # mean among observations with imputed values
  Xbar <- colMeans(X[m==1,,drop=FALSE])

  # observations without imputed values
  X <- X[m==0,,drop=FALSE]
  Xk <- ncol(X)

  X_hasvar <- apply(X, 2, var)!=0 
  X <- X[,X_hasvar,drop=FALSE]
  Xbar <- Xbar[X_hasvar]

  if(verbose>0){
    message("Formula: ", formula_string(formula) )
    message("Number of covariates: ", Xk )
    message("Covariates w/o variation: ", sum(!X_hasvar) )
    message("Final number of covariates: ", ncol(X))
    }

  return(list(X=X,Xbar=Xbar))
  }

construct_m <- function(df,formula){
  yname <- all.vars(formula)[1]
  return(as.numeric(is.na(df[,yname])))
  }


tolower_remove_dot <- function(x){
  return(gsub("\\.","_",tolower(x)))
  }

tolower_remove_dot_vars <- function(df){
  colnames(df) <- tolower_remove_dot(colnames(df))
  return(df)
}


make_categorical_cov <- function(
  original_df, 
  imputed_df, 
  formula, 
  categories, 
  convert){

  is_numeric <- sapply(original_df, is_var_type, type=convert)

  has_na <- apply(is.na(original_df), 2, sum)>0
  num_vars <- names(is_numeric)[is_numeric==TRUE & has_na==TRUE]
  na_vars <- names(has_na)[has_na==TRUE]

  yname <- all.vars(formula)[1]

  # Only retain imputations for y
  tmp <- original_df
  tmp[[yname]] <- imputed_df[[yname]]

  # Numeric variables to categorical
  for(num_var in num_vars){
    if(num_var != yname){
      tmp[[num_var]] <- make_categorical(tmp[[num_var]],C=categories+1)
    }
  }

  # Explicat the missing values 
  for(na_var in na_vars){
    if(na_var != yname){
      tmp[[na_var]] <- explicat_na(tmp[[na_var]])
    }
  }
  
  return(tmp)
  }


#' @importFrom stats quantile
make_categorical <- function(x, C){
  q <- quantile(x, probs=seq(0,1,length=C), na.rm=TRUE)
  return(cut(x, breaks=q, ordered_result=TRUE, include.lowest=TRUE))
  }

# Check if x is of type "type"
is_var_type <- function(x, type){
  return( sum(class(x) %in% type)>0 )
}

# Transform NA in new category
explicat_na <- function(x, na_level="(Missing)" ){
    if( !is.factor(x) ) {
      x <- factor(x)
    }
    is_missing <- is.na(x)
    if (any(is_missing)) {
        levels(x) <- c(levels(x), na_level)
        x[is_missing] <- na_level
    }
    return(x)
}


R/anes08.R
#' 11 Variables from the 2008 American National Election Study
#'
#' The survey dataset includes 11 variables from the 
#' 2008 edition of the American National Election Study. 
#'
#' @format A data frame with 2265 rows and 11 variables:
#' \describe{
#'  \item{age}{Age of the respondent}
#'  \item{female}{Sex of the respondent}
#'  \item{white}{Non-white vs. white}
#'  \item{education}{No high school, 
#'   some high school, high school diploma, college}
#'  \item{income}{Low, medium, high.}
#'  \item{religion}{Protestant, Catholic/Orthodox, Atheist/other}
#'  \item{married}{Single, married, no longer married}
#'  \item{jobs_r}{7 response categories ranging from 
#'  "Govt should let each person get ahead on own" (1) 
#'  to "Govt should see to jobs and standard of living" (7)}
#' \item{imp_enviro}{Whether the respondent sees the environment 
#'  as an important issue (important/not important)}
#' \item{vote}{ Vote choice in the 2008 presidential 
#'  election: Obama, McCain, No vote/Other}
#' \item{time}{Time for the respondent to complete the survey}
#' }
#' 
#' @details 
#' The dataset \code{anes} was prepared by Kropko et al. (2014). The nine 
#' variables contain different levels of missingness. 
#' 
#' The dataset was imputed using random value imputation and predictive 
#' mean matching via [mice::mice()] (Version: 3.13.0, seed=42).
#' 
#' The imputed datasets are available in \code{anes_rng} (random value
#' imputation) and \code{anes_pmm} (predictive mean matching).
#' 
#' @aliases anes08_rng anes08_pmm
#' 
#' @references
#' J. Kropko, B. Goodrich, A. Gelman, and J. Hill. 2014. Multiple Imputation 
#' for Continuous and Categorical Data: Comparing Joint Multivariate Normal 
#' and Conditional Approaches, Political Analysis 22(4):497–519.
#' 
#' @source \url{https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/24672}
"anes08"

R/wrapper.R
run_sbw <- function(
    original,
    imputed,
    formula,
  bal_alg,
  bal_tol,
  bal_std,
  bal_gri,
  bal_sam,
  sol_nam,
  skip_n,
  scale,
  verbose,
  fit.obj=FALSE){

  m <- construct_m(df=original,formula=formula)

  if(sum(m)<skip_n) {
   if(verbose>0){
      message("Formula: ", formula_string(formula) )
      message("Number of imputed values: ", sum(m) )
      message("Skipping." )
      }
    return(NULL)
   }

  X_Xbar <- construct_x(df=imputed,formula=formula,m=m,scale=scale,verbose=verbose)
  X <- X_Xbar[['X']]
  Xbar <- X_Xbar[['Xbar']]
  
  Xbar_names <- names(Xbar)
  X <- as.data.frame(X)

  sbw_out <- sbw::sbw(dat=X,
               bal=list(
                bal_cov=Xbar_names,
                bal_alg=bal_alg,
                bal_tol=bal_tol,
                bal_std=bal_std,
                bal_gri=bal_gri,
                bal_sam=bal_sam),
               wei = list(
                wei_sum=TRUE, 
                wei_pos=TRUE),
               sol=list(
                sol_nam=sol_nam,
                sol_dis=(verbose>1)),
               par=list(
                par_est="aux", 
                par_tar=Xbar))

  if(verbose>1){
    sbw::summarize(sbw_out)
  }

  if(fit.obj){
    return(sbw_out)
  } else {
    w <- rep(NA,length(m))
    w[m==0] <- sbw_out$dat_weights$sbw_weights*sum(m)
    w[m==1] <- 1
    return(w) 
    }


  }


run_ebal <- function(
    original,
    imputed,
    formula,
  coefs,
  max.iterations,
  base.weight,
  constraint.tolerance,
  norm.constant,
  max.weight,
  min.weight,
  max.trim.iterations,
  max.weight.increment,
  min.weight.increment,
  trim, 
  skip_n,
  scale,
  verbose,
  fit.obj=FALSE){

  m <- construct_m(df=original,formula=formula)

  if(sum(m)<skip_n) {
   if(verbose>1){
      message("Formula: ", formula_string(formula) )
      message("Number of imputed values: ", sum(m) )
      message("Skipping." )
      }
    return(NULL)
   }

  X_Xbar <- construct_x(df=imputed,formula=formula,m=m,scale=scale,verbose=verbose)
  X <- X_Xbar[['X']]
  Xbar <- X_Xbar[['Xbar']]

  # Modified code from ebal::ebalance() >>>
 
  if (length(max.iterations) != 1 ) {
    stop("length(max.iterations) != 1")
    }

  if (length(constraint.tolerance) != 1 ) {
    stop("length(constraint.tolerance) != 1")
    }
    
  # set up elements
  ntreated  <- sum(m==1)
  ncontrols <- sum(m==0)

  if (is.null(base.weight)) {
    base.weight = rep(1, ncontrols)
    }

  if ( length(base.weight) !=  ncontrols) {
    stop("length(base.weight) !=  number of controls  sum(Treatment==0)")
    }


  #co.x <- X[Treatment==0,]
  #co.x <- cbind(rep(1,ncontrols),co.x)

  co.x <- X
  co.x <- cbind(rep(1,ncontrols),co.x)

  if(qr(co.x)$rank != ncol(co.x)){
    stop("collinearity in covariate matrix for controls (remove collinear covariates)")
    }

  #tr.total <- apply(as.matrix(X[Treatment==1,]),2,sum)
  tr.total <- Xbar*ntreated

  if (is.null(norm.constant)) {
    norm.constant <- ntreated
    }

  if (length(norm.constant) != 1) {
    stop("length(norm.constant) != 1")
    }

  tr.total <- c(norm.constant,tr.total)

  if(is.null(coefs)) {
    coefs = c(log(tr.total[1]/sum(base.weight)),rep(0,(ncol(co.x)-1)))
    }
   
  if(length(coefs) != ncol(co.x)) {
    stop("coefs needs to have same length as number of covariates plus one")
    }
   
  eb.out <- ebal::eb(tr.total=tr.total,
               co.x=co.x,
               coefs=coefs,
               base.weight=base.weight,
               max.iterations=max.iterations,
               constraint.tolerance=constraint.tolerance,
               print.level=verbose)

  ebal_out <- list(
            target.margins = tr.total,
            co.xdata = co.x,
            w=eb.out$Weights.ebal,
            coefs=eb.out$coefs,
            maxdiff=eb.out$maxdiff,
            norm.constant = norm.constant,
            constraint.tolerance=constraint.tolerance,
            max.iterations=max.iterations,
            base.weight=base.weight,
            print.level=verbose,
            converged=eb.out$converged
      )
        
  class(ebal_out) <- "ebalance"

  # <<< end 

  if(trim){

    ebal_out <- ebal::ebalance.trim(ebal_out, 
      max.weight = max.weight,
      min.weight = min.weight, 
      max.trim.iterations = max.trim.iterations,
      max.weight.increment = max.weight.increment,
      min.weight.increment = min.weight.increment,
      print.level = verbose)

  }

  if(fit.obj){
    return(ebal_out)
  } else {
    w <- rep(NA,length(m))
    w[m==0] <- ebal_out$w
    w[m==1] <- 1
    return(w) 
    }
  }






data-raw/anes.R
library(foreign)
library(mice)

# Download data from: https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/24672

anes <- foreign::read.dta("./rawdata/base2008_3.dta")
anes[["jobs_r"]] <- unclass(anes[["jobs_r"]])

##### CODE FROM Kropko et. al 
# BEGIN >>> 
anes <- anes[ ,c(6:8, 11:13, 25, 42, 44:46)]

levels(anes$imp_enviro) <- c( "Not important" , "Not important", "Important", "Important" )

levels(anes$religion) <- c("Protestant", "Catholic/Orthodox", "Atheist/Other", "Atheist/Other")
anes <- anes[!as.logical(rowSums(is.na(anes[,c(2,3,9,11)]))),] 

anes[["education"]] <- factor(anes[["education"]], ordered=TRUE)
# anes[["jobs_r"]] <- factor(anes[["jobs_r"]], ordered=TRUE)
anes[["income"]] <- factor(anes[["income"]], ordered=TRUE)
# <<< END 

anes[['time']] <- log(anes[['time']])

vars <- c("age", "female", 
	"white", "education", "income", "religion", 
	"married", "jobs_r", "imp_enviro", "vote", "time")

anes08 <- anes[, vars]


# Impute 
#########

anes08_rng <- mice(anes08, method='sample', seed=42)
anes08_pmm <- mice(anes08, method='pmm', seed=42)


usethis::use_data(anes08, overwrite = TRUE)
usethis::use_data(anes08_rng, overwrite = TRUE)
usethis::use_data(anes08_pmm, overwrite = TRUE)


R/missDiag.R
#' Comparing Observed and Imputed Values under MAR and MCAR
#'
#' Function to compute discrepancy statistics comparing the (reweighted) 
#' density of imputed and observed values. To compute the weights balancing
#' covariates that are part of the missing data mechanism, the function 
#' relies on either entropy balancing or stable balancing weights.
#' 
#' @param original data frame with missing values. 
#' @param imputed an imputed data frame, a list of imputed data frames or 
#'  a \code{mids} object from the [mice::mice()] package.
#' @param formula two-sided formula with the variables to compute discrepancy
#'  statistics (right-hand side) and variables to balance (left-hand side). 
#' @param skip_n skip variables with fewer than \code{skip_n} missing values. 
#' @param scale scale the design matrix?
#' @param adjust either 'none', 'sbw' or 'ebal'.
#' @param use_imputed_cov use imputed covariates? 
#' @param convert vector of variable types to bin before coercion
#'  (ignored if \code{use_imputed_cov=TRUE}).
#' @param categories how many bins (ignored if \code{use_imputed_cov=TRUE}).
#' @param verbose one of three levels (0,1 or 2). 
#' @param output_diag add covariate balance statistics to output? 
#' @param cobalt_param list of parameters passed to \code{cobalt}.
#' @param ebal_param list of parameters passed to \code{ebal}.
#' @param sbw_param list of parameters passed to \code{sbw}.
#'   
#' @details 
#' Let y be a variable with observed and imputed values and let m be 
#' a corresponding indicator if a value in y is observed (0) or imputed (1). 
#' We use X to denote all K covariates and x_k as the k^th covariate. 
#' 
#' By default, \code{missDiag} computes discrepancy statistics (balance
#' statistics) comparing the distribution of observed and imputed values, 
#' ie. comparing f(y|m=1) vs. f(y|m=0). The distribution are expected 
#' to be equal under MCAR. 
#' 
#' If the data are missing at random (MAR), set \code{adjust="ebal"} 
#' or \code{adjust="sbw"}, to construct weights to balance the 
#' covariate distribution X before computing the discrepancy statistics. 
#' 
#' The left-hand side (lhs) of the \code{formula} gives the name of y while the 
#' right-hand side (rhs) defines the covariates X to balance under MAR.
#' Rhs variables are transformed into a numeric matrix via the 
#' [stats::model.matrix()] function. Therefore  factor and string variables 
#' are dummy encoded and functional transformations such as \code{log()} 
#' are applied. 
#' 
#' To run \code{missDiag} on many y variables sequentially, supply a list 
#' of formulas or name all y variables on the lhs separating each variable 
#' with a "|", e.g. \code{y1 | y2 ~ x1 + ... + xK}.
#' 
#' If \code{formula=NULL}, \code{missDiag} runs on all variables with at 
#' least \code{n_skip} missing values in \code{original}. Weights are 
#' constructed such that all X covariates are balanced 
#' (if \code{adjust!='none'}). 
#' 
#' If \code{use_imputed_cov=FALSE}, all imputed values in the covariates X 
#' are deleted, variables are coerced to factor variables with missing 
#' values encoded as one category. Numerical variables are binned by quantiles 
#' before coercion. Use the parameter \code{categories} to define the 
#' number of bins. By default only variables of class "numeric" are binned. 
#' Use \code{convert = c("numeric", "integer")} to also bin integer 
#' variables before coercion to factor variables. 
#' 
#' To balance the covariates, \code{missDiag} computes weights using either 
#' the [ebal::ebalance()] function or the [sbw::sbw()] function. These 
#' two packages have to be installed separately by the user. 
#'  
#' To pass parameters to the [sbw::sbw()] or [ebal::ebalance()] use the 
#' functions [param_sbw()] and [param_sbw()] to generate valid parameter lists 
#' and pass these lists via the parameters \code{sbw_param} or 
#' \code{ebal_param}. 
#' 
#' To display information about the computations set verbose to the value 
#' 1 (some information) or 2 (more information when constructing weights). 
#' 
#' To compute the discrepancy statistics, \code{missDiag} relies on the function
#' [cobalt::bal.tab()] from the cobalt package. To pass parameters to this 
#' function use [param_cobalt()] to generate a valid parameter list and 
#' pass this list via the parameter \code{cobalt_param}.
#' 
#' 
#' @return 
#' A single \code{data.frame} with the results for 
#' each imputed dataset and all lhs variables. 
#' 
#' If \code{output_diag=FALSE}, the dataset only includes 
#' the discrepancy statistics comparing y's observed values and 
#' imputed values.
#' 
#' If y is continuous, there will be exactly one row per imputed 
#' dataset. The discrepancy statistics are reported in the 
#' respective columns. For a categorical y cobalt 
#' computes discrepancy statistics for all K categories which means 
#' that there will be K rows for each imputed dataset in the 
#' output. The label for each category is listed in \code{vname}. 
#' 
#' If the default settings are adopted, the dataset includes
#' the following discrepancy statistics for continuous variables: 
#' 
#' \itemize{
#'  \item \code{diff_adj}: Standardized mean difference
#'  \item \code{v_ratio_adj}: Variance ratio 
#'  \item \code{ks_adj}: Kolmogorov–Smirnov statistic
#'  \item \code{ovl_adj}: 1-Overlap coefficient 
#' }
#' 
#' No variance ratio is reported for categorical variables.
#' 
#' If \code{output_diag=TRUE}, the dataset also includes 
#' the balance statistics for all K covariates. 
#' 
#' These balance statistics are useful to diagnose if the fitted 
#' weights are successful in balancing the covariate distribution.
#' 
#' For these diagnostics the column \code{vname} lists 
#' the covariate name (for continuous variables) or the 
#' covariate category (for categorical variables). 
#' 
#' @references
#' Moritz Marbach. 2021. Choosing Imputation Models.
#' 
#' José R Zubizarreta. 2015. Stable Weights that Balance Covariates 
#' for Estimation with Incomplete Outcome Data, Journal of the American 
#' Statistical Association, 110(511): 910-922.
#' 
#' Jens Hainmueller. 2012. Entropy Balancing for Causal Effects: A 
#' Multivariate Reweighting Method to Produce Balanced Samples in 
#' Observational Studies, Political Analysis 20(1): 25–46.
#' 
#' @examples 
#' 
#' # Compare random value imputation 
#' # with predictive mean imputation 
#' # under MCAR
#' 
#' diag_rng <- missDiag( 
#'  original=anes08, 
#'  imputed=anes08_rng, 
#'  verbose = 1,
#'  adjust = 'none',
#'  formula = time ~ .)
#' 
#' diag_pmm <- missDiag( 
#'  original=anes08, 
#'  imputed=anes08_pmm, 
#'  verbose = 1,
#'  adjust = 'none',
#'  formula = time ~ .)
#' 
#' # SMD: 
#' 
#'  mean(diag_pmm$diff_adj)
#'  mean(diag_rng$diff_adj)
#' 
#' # log(Variance ratio)
#'  mean(log(diag_pmm$v_ratio_adj))
#'  mean(log(diag_rng$v_ratio_adj))
#' 
#' # KS
#'  mean(diag_pmm$ks_adj)
#'  mean(diag_rng$ks_adj)
#'
#' # 1-OVL
#'  mean(diag_pmm$ovl_adj)
#'  mean(diag_rng$ovl_adj)
#'
#' 
#' 
#' @importFrom cobalt bal.tab
#' @importFrom stats as.formula formula model.matrix update.formula var
#' @export
missDiag <- function(
  original, 
  imputed, 
  formula=NULL, 
  skip_n=25,
  scale=FALSE,
  adjust='ebal',
  use_imputed_cov = TRUE,
  convert = c("numeric"),
  categories = 3, 
  verbose=0,
  output_diag=FALSE, 
  ebal_param=param_ebal(),
  sbw_param=param_sbw(),
  cobalt_param=param_cobalt()){

  call_ <- sys.call()

  if(!adjust %in% c("none", "sbw", "ebal")) { 
    stop("Parameter 'adjust' must be one of these: none, sbw or ebal") }

  # Preprocessing: Data 
  original <- as.data.frame(original)

  if("mids" %in% class(imputed)){
  
    imputed <- mice::complete(imputed,action='all')
    M <- length(imputed)
  
  } else if("list" %in% class(imputed)) {

    M <- length(imputed)

  } else {

    imputed <- list(as.data.frame(imputed))
    M <- 1

  }

  # Preprocessing: Formula 
  if(!is.null(formula)){

    formulas <- construct_formulas(formula, 
      vars=colnames(original))

  } else {

    vars <- colnames(original)
    vars_na <- apply(is.na(original), 2, sum)>0
    vars <- vars[vars_na]

    dv <- paste(vars, collapse = "|")
    iv <- paste(colnames(original), collapse="+")
    formula <- as.formula(paste0(dv, "~", iv))
    formulas <- construct_formulas(formula, vars=vars)

    }


  out <- list()

  for(m in 1:M){

    if(verbose>0 & M>1) cat("Dataset Nummer:",m,"\n")

    imp <- as.data.frame(imputed[[m]])

    for(f in formulas){

      # Reprocessing X 
      if( use_imputed_cov == FALSE){

        imp_prep <- make_categorical_cov(
           original_df = original,  
           imputed_df = imp, 
           formula = as.formula(f), 
           categories = categories,
           convert = convert)

      } else {

        imp_prep <- imp 

      }

      # Construct weights      
      if(adjust=='sbw'){
        w <- do.call(run_sbw, c(
            list(original= original,
                 imputed = imp_prep, 
                 formula = as.formula(f),
                 skip_n  = skip_n, 
                 scale = scale,
                 verbose = verbose), 
            sbw_param))
      }

      if(adjust=='ebal'){
        w <- do.call(run_ebal, c(
            list(original= original,
                 imputed = imp_prep, 
                 formula = as.formula(f),
                 skip_n  = skip_n, 
                 scale = scale,
                 verbose = verbose), 
            ebal_param))
      }

      if(adjust=='none'){
        if(verbose>0){
          message("No weights constructed." )
        }

        w <- rep(1,nrow(imp_prep))
      }
      
      if(!is.null(w)){

        # Balance 
        yname <- all.vars(f)[1]
        xname <- all.vars(f)

        res <- do.call(bal.tab, c(
            list(x=imp_prep[,xname,drop=FALSE],
                 treat = is.na(original[[yname]]), 
                 weights = w), 
            cobalt_param))$Balance

        res$m <- m 
        res$vname_lhs <- yname
        res$vname <- rownames(res)

        out[[length(out)+1]] <- res

      }

    }
  }

  out <- do.call(rbind,out) 
  rownames(out) <- NULL

  out <- tolower_remove_dot_vars(out)
  out <- out[,!grepl("_un|_threshold", colnames(out))]
  out$diag <- ifelse(mapply(grepl, out$vname_lhs, out$vname), 
    FALSE, TRUE)

  attr(out, 'call') <- call_

  if(output_diag==FALSE){
    out <- subset(out, diag==FALSE)
  }

  return(out)
  }
